x-pack/filebeat/input/httpjson/request.go

```diff
@@ -163,21 +181,160 @@ func newRequester(
 	}
-	defer httpResp.Body.Close()
+	return httpResp, nil
+}
 
-	eventsCh, err := r.responseProcessor.startProcessing(stdCtx, trCtx, httpResp)
+// generateNewUrl returns new url value using replacement from oldUrl with str ids
+func generateNewUrl(replacement string, oldUrl string, str []byte) (url.URL, error) {
+	reg, err := regexp.Compile(replacement)
+	if err != nil {
+		return url.URL{}, fmt.Errorf("failed to create regex on provided value: %w", err)
+	}
+	newUrl, err := url.Parse(string(reg.ReplaceAll([]byte(oldUrl), str)))
 	if err != nil {
-		return err
+		return url.URL{}, fmt.Errorf("failed to replace value in url: %w", err)
+	}
+	return *newUrl, nil
+
+}
+
+func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher) error {
+	var (
+		n                 int
+		ids               []string
+		err               error
+		split             *split
+		urlCopy           url.URL
+		urlString         string
+		httpResp          *http.Response
+		intermediateResps []*http.Response
+		finalResps        []*http.Response
+	)
+	for i, rf := range r.requestFactory {
+		// iterate over collected ids from last response
+		if i != 0 {
+			if len(ids) == 0 {
+				n = 0
+				continue
+			}
+			urlCopy = rf.url
+			urlString = rf.url.String()
+			// perform request over collected ids
+			for _, s := range ids {
+				// reformat urls of requestFactory using ids
+				rf.url, err = generateNewUrl(rf.replace, urlString, []byte(s))
+				if err != nil {
+					return fmt.Errorf("failed to generate new URL: %w", err)
+				}
+
+				// collect data from new urls
+				httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+				if err != nil {
+					return fmt.Errorf("failed to execute http client.Do: %w", err)
+				}
+				// store data according to response type
+				if i+1 == len(r.requestFactory) && len(ids) != 0 {
+					finalResps = append(finalResps, httpResp)
+				} else {
+					intermediateResps = append(intermediateResps, httpResp)
+				}
+			}
+			rf.url = urlCopy
+
+			if i+1 == len(r.requestFactory) {
+				split = r.responseProcessor.split
+				r.responseProcessor.split = nil
+				n, err = r.processAndPublishEvents(0, stdCtx, trCtx, publisher, finalResps)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
+				continue
+			} else {
+				ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactory[i+1].replace)
+				if err != nil {
+					return err
+				}
+				split = r.responseProcessor.split
+				r.responseProcessor.split = nil
+				n, err = r.processAndPublishEvents(1, stdCtx, trCtx, publisher, intermediateResps)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
+			}
+		} else {
+			// perform and store regular call responses
+			httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+			if err != nil {
+				return fmt.Errorf("failed to execute http client.Do: %w", err)
+			}
+			if len(r.requestFactory) <= 1 {
```

Comment:
`len(r.requestFactory)` is never `0` here, so this can be simplified:
```suggestion
			if len(r.requestFactory) == 1 {
```